packages/@atjson/peritext/src/index.ts

```diff
@@ -0,0 +1,585 @@
+    let id = (markCounter++).toString(16);
+    id = `M${"00000000".slice(id.length) + id}`;
+    ids.set(mark.id, id);
+  }
+  for (const block of doc.blocks) {
+    const id = ids.get(block.id);
+    if (id) {
+      block.id = id;
+    }
+    block.attributes = attributesWithStableIds(block.attributes, ids);
+  }
+  for (const mark of doc.marks) {
+    const id = ids.get(mark.id);
+    if (id) {
+      mark.id = id;
+    }
+    mark.attributes = attributesWithStableIds(mark.attributes, ids);
+  }
+
+  return doc;
+}
+
+function unsafe_parseRange(range: Range) {
+  const [lhs, rhs] = range.split("..");
+  const leadingChar = lhs[0];
+  const trailingChar = rhs[rhs.length - 1];
+  const start = parseInt(lhs.slice(1), 10);
+  const end = parseInt(rhs.slice(0, -1), 10);
+
+  return {
+    start,
+    end,
+    edgeBehaviour: {
+      leading:
+        leadingChar === "(" ? EdgeBehaviour.preserve : EdgeBehaviour.modify,
+      trailing:
+        trailingChar === ")" ? EdgeBehaviour.preserve : EdgeBehaviour.modify,
+    },
+  };
+}
+
+function serializeRange(
+  start: number,
+  end: number,
+  edgeBehaviour: { leading: EdgeBehaviour; trailing: EdgeBehaviour }
+): Range {
+  return `${
+    edgeBehaviour.leading === EdgeBehaviour.preserve ? "(" : "["
+  }${start}..${end}${
+    edgeBehaviour.trailing === EdgeBehaviour.preserve ? ")" : "]"
+  }`;
+}
+
+function adjustMarkRange(mark: Mark, offset: number): Mark {
+  const { start, end, edgeBehaviour } = unsafe_parseRange(mark.range);
+
+  return {
+    ...mark,
+    range: serializeRange(start + offset, end + offset, edgeBehaviour),
+  };
+}
+
+function adjustMarkEnd(mark: Mark, offset: number): Mark {
+  const { start, end, edgeBehaviour } = unsafe_parseRange(mark.range);
+
+  return {
+    ...mark,
+    range: serializeRange(start, end + offset, edgeBehaviour),
+  };
+}
+
+/**
+ * @class PeritextBuilderStep is a peritext document, suitable
+ *   for use with other peritext functions, but with an additional property:
+ * @property value - contains the direct result of a peritext builder function
+ *   so that you can easily access the newly generated ID (for instance)
+ */
+class PeritextBuilderStep<ReturnT> {
+  constructor(
+    public text: string,
+    public blocks: Block[],
+    public marks: Mark[],
+    private value: ReturnT
+  ) {}
+
+  static fromPeritext<ReturnT>(
+    doc: Peritext,
+    value: ReturnT
+  ): PeritextBuilderStep<ReturnT> {
+    return new PeritextBuilderStep(doc.text, doc.blocks, doc.marks, value);
+  }
+
+  /**
+   * Strips the `value` field off of the data, so the resulting document
+   *   can be cleanly shared;
+   * @returns the equivalent peritext document, minus the `value` field
+   */
+  public peritext(): Peritext {
+    const { blocks, marks, text } = this;
+    return { blocks, marks, text };
+  }
+
+  public getValue(): ReturnT {
+    return this.value;
+  }
+
+  public set<K extends keyof ReturnT>(key: K, value: ReturnT[K]) {
```

Comment:
somewhat surprising to me that we would want to expose this setter. What is the use case?

Reply:
Oh nvm, was misreading. I thought this was setting the `value`. This allows someone to set attributes _on_ the value block or mark. This is 👍

Reply:
yeah, the only use case I have for it presently is setting the `selfClosing` property on blocks, but it would also let you set the attributes in-line if you wanted, or you could set the IDs manually. I don't think there's a like...valid reason to set the parents on a block or the range on a mark but like I don't see a reason to go out of our way to prevent that here